src/tabs/map.rs

```diff
@@ -837,18 +919,90 @@ impl tab::Tab for Tab {
                     {
                         self.dragging_event = false;
                         self.event_drag_offset = None;
-                        self.add_event(&mut map);
+                        if let Some(id) = self.add_event(&mut map) {
+                            self.redo_history.clear();
+                            if self.history.len() == HISTORY_SIZE {
+                                self.history.pop_front();
+                            }
+                            self.history.push_back(HistoryEntry::EventCreated(id));
+                        }
+                    }
+                }
+
+                // Handle undo/redo keypresses
+                let is_dragged_by_primary = response.dragged_by(egui::PointerButton::Primary);
+                let is_undo_pressed = ui.input(|i| {
+                    i.modifiers.command && !i.modifiers.shift && i.key_pressed(egui::Key::Z)
+                });
+                let is_redo_pressed = ui.input(|i| {
+                    i.modifiers.command
+                        && (i.modifiers.shift || i.key_pressed(egui::Key::Y))
+                        && (!i.modifiers.shift || i.key_pressed(egui::Key::Z))
+                });
+                if !is_dragged_by_primary && (is_undo_pressed || is_redo_pressed) {
+                    let new_entry = match if is_undo_pressed {
+                        self.history.pop_back()
+                    } else {
+                        self.redo_history.pop()
+                    } {
+                        None => None,
+
+                        Some(HistoryEntry::Tiles { layer, mut delta }) => {
+                            for d in delta.iter_mut() {
+                                let position = (d.0, d.1, layer);
+                                let new_id = d.2;
+                                *d = (d.0, d.1, map.data[position]);
+                                map.data[position] = new_id;
+                                self.view.map.set_tile(new_id, position);
+                            }
+                            Some(HistoryEntry::Tiles { layer, delta })
+                        }
+
+                        Some(HistoryEntry::EventMoved { id, x, y }) => {
+                            let event = map.events.get_mut(id).unwrap();
+                            let new_entry = Some(HistoryEntry::EventMoved {
+                                id,
+                                x: event.x,
+                                y: event.y,
+                            });
+                            event.x = x;
+                            event.y = y;
+                            new_entry
+                        }
+
+                        Some(HistoryEntry::EventCreated(id)) => {
+                            let event = map.events.remove(id);
```

Comment:
Double checking, this doesn't remove an event window if this event is open. 
Fixing this would take rewriting the `Windows` struct to take a generic parameter like `Tabs`, and I'll be doing that in the code refactor, so you don't need to fix this right now